common/expression/src/types/string.rs

```diff
@@ -258,4 +258,61 @@ impl StringColumnBuilder {
         assert_eq!(self.offsets.len(), 2);
         self.data[(self.offsets[0] as usize)..(self.offsets[1] as usize)].to_vec()
     }
+
+    pub fn try_from_transform<F>(
+        src: StringIterator<'_>,
+        estimate_bytes: usize,
+        mut f: F,
+    ) -> Result<StringColumnBuilder, String>
+    where
+        F: FnMut(&[u8], &mut [u8]) -> Result<usize, String>,
+    {
+        let mut values: Vec<u8> = Vec::with_capacity(estimate_bytes);
+        let mut offsets: Vec<u64> = Vec::with_capacity(src.size_hint().0 + 1);
+        offsets.push(0);
+
+        let mut offset: usize = 0;
+        unsafe {
+            for x in src {
+                let bytes = std::slice::from_raw_parts_mut(
+                    values.as_mut_ptr().add(offset),
+                    values.capacity() - offset,
+                );
+
+                match f(x, bytes) {
+                    Ok(l) => {
+                        offset += l;
+                        offsets.push(offset as u64);
+                    }
+
+                    Err(e) => return Err(e),
+                }
+            }
+            values.set_len(offset);
+            values.shrink_to_fit();
+
+            Ok(StringColumnBuilder {
+                data: values,
+                offsets,
+            })
+        }
+    }
+}
+
+pub fn try_transform_scalar<F>(
+    val: &[u8],
+    estimate_bytes: usize,
+    mut func: F,
+) -> Result<Vec<u8>, String>
+where
+    F: FnMut(&[u8], &mut [u8]) -> Result<usize, String>,
+{
+    let mut buf = Vec::with_capacity(estimate_bytes);
+    unsafe {
+        let bytes = std::slice::from_raw_parts_mut(buf.as_mut_ptr(), buf.capacity());
```

Comment:
As the doc of `std::slice::from_raw_parts_mut` say:
```
/// * `data` must be [valid] for both reads and writes for `len * mem::size_of::<T>()` many bytes,
```
It may result in UB here.

Do we need use `vec![0u8; estimate_bytes]` here?

Reply:
How about using:

`let mut buf = vec![0; estimate_bytes] `

Reply:
As the doc of `std::slice::from_raw_parts_mut` says:
```
/// * `data` must be [valid] for both reads and writes for `len * mem::size_of::<T>()` many bytes,
```
It may result in UB here.

can we use `vec![0u8; estimate_bytes]`?

Reply:
I'd like not to optimize allocation for String scalar. If we reach an agreement, we can remove this function totally.

Reply:
It is ok to me to remove this optimize for String scalar.But we still need to resolve UB for String Column.

https://github.com/ygf11/databend/blob/ea69d75871aa30624ed24d2d60bc6b9bdcc7394d/common/expression/src/types/string.rs#L270-L279